tsdb/load_shards.go

```diff
@@ -0,0 +1,350 @@
+package tsdb
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/influxdata/influxdb/logger"
+	"github.com/influxdata/influxdb/pkg/limiter"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// res holds the result from opening each shard in a goroutine
+type res struct {
+	s   *Shard
+	err error
+}
+
+type shardLoadingContext struct {
+	db    os.DirEntry
+	rp    os.DirEntry
+	sh    os.DirEntry
+	sFile *SeriesFile
+	resC  chan *res
+	idx   interface{}
+	t     limiter.Fixed
+	wg    *sync.WaitGroup
+	log   *zap.Logger
+}
+
+func (s *Store) loadShards() error {
+	err, shardCtx, dbDirs := s.setupShardLoader()
+	if err != nil {
+		return err
+	}
```

Comment:
Not sure it makes sense to pull code out of `loadShards` into methods like `setupShardLoader`. It doesn't seem to improve readability, at least for me, and the refactor doesn't reduce duplicate code. It's also created some issues. When pulling code out of function you have to watch `defer` statements. At least one `defer` was pulled out, the one that ends the logging operation for `tsdb_open`. This means the log will show `tsdb_open` completed after `setupShardLoader` finishes, instead of when `loadShards` finishes. That _could_ be fixed and keep `setupShardLoader` intact, but it would then make both `loadShards` and `setupShardLoader` more complicated.